### DISCUSSION TITLE
如何使用 AWS s3 协议连接 backblaze b2

### DISCUSSION THREAD

**Discussion by 0xxb** [lsky-org/lsky-pro#448] [教程]
实际上兰空图床是支持 backblaze b2 的，只不过 AWS s3 的 sdk 中，默认存在 ACL(https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html) 配置，参数值默认为 private，因为 b2 控制台中没有 ACL 设置，使用公开的 bucket 会出现错误：`Cannot parse S3 Canned ACL value of 'private'`，所以需要稍微修改点代码，覆盖 acl  配置。

打开文件：https://github.com/lsky-org/lsky-pro/blob/a8ab0c6680eaac6b767286525e2c1b4cbd14bcba/app/Services/ImageService.php#L339

在 339 行下面追加一行：
```php
options: ['params' => ['ACL' => '']],
```

示例：
```php
StrategyKey::Minio => new AwsS3V3Adapter(
    // ...
    bucket: $configs->get(MinioOption::Bucket),
    options: ['params' => ['ACL' => '']],  // <-- 这里是新增的一行
    // ...
),
```

然后在创建储存策略时，选择 Minio 进行配置即可，以下是配置示例。
![image](https://user-images.githubusercontent.com/22728201/173543674-5ed51652-277e-4a42-b94e-aa5c4c8f1eb9.png)

其中的 Bucket、连接地址和区域，可以在 b2 控制台创建 Backet 以后找到，AccessKey 和 SecretKey 便是控制台中 App Key 中创建的前面，key ID = AccessKey，applicationKey = Secret Key。

注意：
- 不要使用 Master Application Key，它不兼容 S3 协议。
- 在 2020 年 5 月 4 日之前创建的桶不支持 s3 协议。

**Comment by mouyase** [comment]
学习了

**Comment by alonedogboy** [comment]
上传失败, 图片上传失败
弄完之后无法上传

**Comment by smalljunxing** [comment]
解决了，我的在356行。

**Comment by KKISO574** [comment]
可以用这个创建阿里oss吗，我香港的存储桶一直挂不上，不显示，alist都可以挂上的。

**Comment by qq80284445** [comment]
可以直接用AwsS3策略一样的，我试过了，不想大改，就把/app/Services/ImageService.php 里面的StrategyKey::S3 => new AwsS3V3Adapter
改成了这样：
```
StrategyKey::S3 => new AwsS3V3Adapter(
                client: new S3Client([
                    'credentials' => [
                        'key'    => $configs->get(S3Option::AccessKeyId),
                        'secret' => $configs->get(S3Option::SecretAccessKey)
                    ],
                    'endpoint' => $configs->get(S3Option::Endpoint),
                    // 根据条件判断是否去除 @backblaze 并设置 region
                    'region'  => strpos($configs->get(S3Option::Region), '@backblaze') !== false 
                                 ? str_replace('@backblaze', '', $configs->get(S3Option::Region)) 
                                 : $configs->get(S3Option::Region),
                    'version' => '2006-03-01',
                ]),
                bucket: $configs->get(S3Option::Bucket),
                // 根据条件判断是否添加 options
                options: strpos($configs->get(S3Option::Region), '@backblaze') !== false 
                         ? ['params' => ['ACL' => '']] 
                         : null
            ),
```
改成这样也不影响AwsS3正常使用，如果是backblaze空间，就在区域(region)里面填区域@backblaze就好了。
![微信截图_20240803015321](https://github.com/user-attachments/assets/c57fb28e-1419-48c8-8b2b-bd47f7143d1f)
然后修改下/resources/views/admin/strategy 目录下add.blade.php edit.blade.php加个提示，怕忘了
